File: source/plugins/datatypes/converternodes.cpp
#define GLM_FORCE_SWIZZLE
#include "glm/glm.hpp"
#include "data/cache_main.h"
#include "converternodes.h"

using namespace MindTree;

void registerConverteNodeOperators()
{
    auto floatToColor = [](DataCache *cache)
    {
        double r = cache->getData(0).getData<double>();
        double g = cache->getData(1).getData<double>();
        double b = cache->getData(2).getData<double>();
        double a = cache->getData(3).getData<double>();

        cache->pushData(glm::vec4(r, g, b, a));
    };

    auto floatToVector3D = [](DataCache *cache)
    {
        double x = cache->getData(0).getData<double>();
        double y = cache->getData(1).getData<double>();
        double z = cache->getData(2).getData<double>();

        cache->pushData(glm::vec3(x, y, z));
    };

    auto floatToVector2D = [](DataCache *cache)
    {
        double x = cache->getData(0).getData<double>();
        double y = cache->getData(1).getData<double>();

        cache->pushData(glm::vec2(x, y));
    };

    auto vector3DToFloat = [](DataCache *cache)
    {
        glm::vec3 vec = cache->getData(0).getData<glm::vec3>();
        cache->pushData(vec.x);
        cache->pushData(vec.y);
        cache->pushData(vec.z);
    };

    auto vector2DToFloat = [](DataCache *cache)
    {
        glm::vec2 vec = cache->getData(0).getData<glm::vec2>();
        cache->pushData(vec.x);
        cache->pushData(vec.y);
    };

    auto colorToFloat = [](DataCache *cache)
    {
        glm::vec4 vec = cache->getData(0).getData<glm::vec4>();
        cache->pushData(vec.r);
        cache->pushData(vec.g);
        cache->pushData(vec.b);
        cache->pushData(vec.a);
    };

    DataCache::addProcessor(new CacheProcessor("COLOR", "FLOATTOCOLOR", floatToColor));
    DataCache::addProcessor(new CacheProcessor("VECTOR3D", "FLOATTOVECTOR3D", floatToVector3D));
    DataCache::addProcessor(new CacheProcessor("VECTOR2D", "FLOATTOVECTOR2D", floatToVector2D));
    DataCache::addProcessor(new CacheProcessor("FLOAT", "COLORTOFLOAT", colorToFloat));
    DataCache::addProcessor(new CacheProcessor("FLOAT", "VECTOR3DTOFLOAT", vector3DToFloat));
    DataCache::addProcessor(new CacheProcessor("FLOAT", "VECTOR2DTOFLOAT", vector2DToFloat));
}

File: source/plugins/datatypes/converternodes.py
import MT

class FloatToVector3DDecorator(MT.pytypes.NodeDecorator):
    label = "Values.Converters.FloatToVector3D"
    type = "FLOATTOVECTOR3D"
    insockets = [
            ("X", "FLOAT"),
            ("Y", "FLOAT"),
            ("Z", "FLOAT")
            ]
    outsockets = [("Vector", "VECTOR3D")]

class FloatToVector2DDecorator(MT.pytypes.NodeDecorator):
    label = "Values.Converters.FloatToVector2D"
    type = "FLOATTOVECTOR2D"
    insockets = [
            ("X", "FLOAT"),
            ("Y", "FLOAT")
            ]
    outsockets = [("Vector", "VECTOR2D")]

class FloatToColorDecorator(MT.pytypes.NodeDecorator):
    label = "Values.Converters.FloatToColor"
    type = "FLOATTOCOLOR"
    insockets = [
            ("R", "FLOAT"),
            ("G", "FLOAT"),
            ("B", "FLOAT"),
            ("A", "FLOAT")
            ]
    outsockets = [("Color", "COLOR")]

class ColorToFloatDecorator(MT.pytypes.NodeDecorator):
    label = "Values.Converters.ColorToFloat"
    type = "COLORTOFLOAT"
    label = "Values.Converters.FloatToColor"
    type = "FLOATTOCOLOR"
    insockets = [("Color", "COLOR")]
    outsockets = [
            ("R", "FLOAT"),
            ("G", "FLOAT"),
            ("B", "FLOAT"),
            ("A", "FLOAT")
            ]

class Vector3DToFloatDecorator(MT.pytypes.NodeDecorator):
    label = "Values.Converters.Vector3DToFloat"
    type = "VECTOR3DTOFLOAT"
    insockets = [("Vector", "VECTOR3D")]
    outsockets = [
            ("X", "FLOAT"),
            ("Y", "FLOAT"),
            ("Z", "FLOAT")
            ]

class Vector2DToFloatDecorator(MT.pytypes.NodeDecorator):
    label = "Values.Converters.Vector2DToFloat"
    type = "VECTOR2DTOFLOAT"
    insockets = [("Vector", "VECTOR2D")]
    outsockets = [
            ("X", "FLOAT"),
            ("Y", "FLOAT")
            ]

MT.registerNode(FloatToVector3DDecorator)
MT.registerNode(FloatToVector2DDecorator)
MT.registerNode(FloatToColorDecorator)
MT.registerNode(Vector3DToFloatDecorator)
MT.registerNode(Vector2DToFloatDecorator)
MT.registerNode(ColorToFloatDecorator)
